semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/cfgdominator/CfgDominatorFrontier.scala

```diff
@@ -2,39 +2,37 @@ package io.shiftleft.semanticcpg.passes.cfgdominator
 
 import scala.collection.mutable
 
+/**
+  * Calculates the dominator frontier for a set of CFG nodes.
+  * The returned multimap associates the frontier nodes to each CFG node.
+  *
+  * The used algorithm is from: "A Simple, Fast Dominance Algorithm" from
+  * "Keith D. Cooper, Timothy J. Harvey, and Ken Kennedy".
+  */
 class CfgDominatorFrontier[NodeType](cfgAdapter: CfgAdapter[NodeType], domTreeAdapter: DomTreeAdapter[NodeType]) {
 
-  /**
-    * Calculates a the dominator frontier for a set of CFG nodes.
-    * The returned multimap associates the frontier nodes to each CFG node.
-    *
-    * The used algorithm is from: "A Simple, Fast Dominance Algorithm" from
-    * "Keith D. Cooper, Timothy J. Harvey, and Ken Kennedy".
-    */
-  def calculate(allCfgNodes: Seq[NodeType]): mutable.MultiDict[NodeType, NodeType] = {
-    val domFrontier = mutable.MultiDict.empty[NodeType, NodeType]
-
-    allCfgNodes.foreach { joinCandiate =>
-      val predecessors = cfgAdapter.predecessors(joinCandiate).iterator.to(Seq)
-
-      if (predecessors.size > 1) {
-        val joinNode = joinCandiate
+  private def doms(x: NodeType): Option[NodeType] = domTreeAdapter.immediateDominator(x)
+  private def pred(x: NodeType): Seq[NodeType] = cfgAdapter.predecessors(x).iterator.to(Seq)
+  private def onlyJoinNodes(x: NodeType): Option[(NodeType, Seq[NodeType])] =
+    Some(pred(x)).filter(_.size > 1).map(p => (x, p))
```

Comment:
Some -> Option: to avoid ending up with `Some(null)`

```suggestion
    Option(pred(x)).collect { 
      case p if p.size > 1 => (x, p)
    }
```

note: be careful applying this, I didn't ask scalac for feedback...